internal/tracinginterceptor/interceptor.go

```diff
@@ -39,32 +45,122 @@ var (
+	}
+	if m.tracer == nil {
+		m.tracer = opentracing.GlobalTracer()
+	}
+	return m
+}
+
+// writer wraps a transport.ResponseWriter to capture additional information for tracing.
+type writer struct {
+	transport.ResponseWriter
+
+	isApplicationError   bool
+	applicationErrorMeta *transport.ApplicationErrorMeta
+	responseSize         int
+}
+
+var _writerPool = sync.Pool{New: func() interface{} {
+	return &writer{}
+}}
+
+func newWriter(rw transport.ResponseWriter) *writer {
+	w := _writerPool.Get().(*writer)
+	*w = writer{ResponseWriter: rw} // reset
+	return w
+}
+
+func (w *writer) SetApplicationError() {
+	w.isApplicationError = true
+	w.ResponseWriter.SetApplicationError()
+}
+
+func (w *writer) SetApplicationErrorMeta(applicationErrorMeta *transport.ApplicationErrorMeta) {
+	if applicationErrorMeta == nil {
+		return
+	}
+
+	w.applicationErrorMeta = applicationErrorMeta
+	if appErrMetaSetter, ok := w.ResponseWriter.(transport.ApplicationErrorMetaSetter); ok {
+		appErrMetaSetter.SetApplicationErrorMeta(applicationErrorMeta)
+	}
+}
+
+func (w *writer) Write(p []byte) (n int, err error) {
+	w.responseSize += len(p)
+	return w.ResponseWriter.Write(p)
+}
+
+func (w *writer) free() {
+	_writerPool.Put(w)
 }
 
 // Handle implements interceptor.UnaryInbound
 func (m *Interceptor) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, h transport.UnaryHandler) error {
-	// TODO: implement
-	panic("implement me")
+	parentSpanCtx, _ := m.tracer.Extract(m.propagationFormat, tchannel.GetPropagationCarrier(req.Headers.Items(), req.Transport))
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
+		ParentSpanContext: parentSpanCtx,
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         CommonTracingTags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	wrappedWriter := newWriter(resw)
+	err := h.Handle(ctx, req, wrappedWriter)
+	if wrappedWriter.isApplicationError {
+		span.SetTag("error.type", "application_error")
+	}
+	wrappedWriter.free()
+
+	return updateSpanWithError(span, err)
 }
 
 // Call implements interceptor.UnaryOutbound
 func (m *Interceptor) Call(ctx context.Context, req *transport.Request, out transport.UnaryOutbound) (*transport.Response, error) {
-	// TODO: implement
-	panic("implement me")
+	createOpenTracingSpan := &transport.CreateOpenTracingSpan{
+		Tracer:        m.tracer,
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     CommonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, tchannel.GetPropagationCarrier(tracingHeaders, m.transport)); err != nil {
+		ext.Error.Set(span, true)
+		span.LogFields(log.String("event", "error"), log.String("message", err.Error()))
+		return nil, err
```

Comment:
I understand that it's existing logic, but should we actually return error for the call because tracing format is wrong?

Reply:
This is a good call, thanks. I updated the code so that the error from m.tracer.Inject is logged and the tracing span is marked as an error, but the RPC call itself will still proceeds.